src/openfl/utils/_internal/ShaderMacro.hx

```diff
@@ -199,6 +436,211 @@ class ShaderMacro
+				return result;
+		}
+	}
+
+	private static function buildGLSLHeaders(glVersion:String):String
+	{
+		if (StringTools.endsWith(glVersion, " compatibility")) return "";
+		if (StringTools.endsWith(glVersion, " core")) return buildGLSLHeaders(StringTools.replace(glVersion, " core", ""));
+
+		return switch (glVersion)
+		{
+			#if desktop
+			case "300 es": "layout (location = 0) out vec4 fragColor;\n";
+			#else
+			case "300 es": "out vec4 fragColor;\n";
+			#end
+
+			case "310 es": buildGLSLHeaders("300 es");
+			case "320 es": buildGLSLHeaders("310 es");
+			case "330": buildGLSLHeaders("320 es");
+			case "400": buildGLSLHeaders("330");
+			case "410": buildGLSLHeaders("400");
+			case "420": buildGLSLHeaders("410");
+			case "430": buildGLSLHeaders("420");
+			case "440": buildGLSLHeaders("430");
+			case "450": buildGLSLHeaders("440");
+			case "460": buildGLSLHeaders("450");
+
+			default: "";
+		};
+	}
+
+	private static function buildGLSLExtensions(glExtensions:Array<{name:String, behavior:String}>, glVersion:String,
+			isFragment:Bool):Array<{name:String, behavior:String}>
+	{
+		if (StringTools.endsWith(glVersion, " compatibility")) return glExtensions;
+		if (StringTools.endsWith(glVersion, " core")) return buildGLSLExtensions(glExtensions, StringTools.replace(glVersion, " core", ""), isFragment);
+
+		switch (glVersion)
+		{
+			default:
+				return glExtensions;
+
+			case "100":
+				return glExtensions;
+			case "110":
+				return glExtensions;
+			case "120":
+				return glExtensions;
+			case "130":
+				return glExtensions;
+			case "140":
+				return glExtensions;
+			case "150":
+				return glExtensions;
+
+			case "300 es":
+				var hasSeparateShaderObjects = false;
+				for (extension in glExtensions)
+				{
+					if (extension.name == "GL_ARB_separate_shader_objects") hasSeparateShaderObjects = true;
+					if (extension.name == "GL_EXT_separate_shader_objects") hasSeparateShaderObjects = true;
+				}
+
+				#if desktop
+				if (!hasSeparateShaderObjects)
+				{
+					#if linux
+					return glExtensions.concat([{name: "GL_EXT_separate_shader_objects", behavior: "require"}]);
+					#else
+					return glExtensions.concat([{name: "GL_ARB_separate_shader_objects", behavior: "require"}]);
+					#end
+				}
+				#end
+				return glExtensions;
+
+			case "310 es":
+				var result = buildGLSLExtensions(glExtensions, "300 es", isFragment);
+				return result;
+
+			case "320 es":
+				var result = buildGLSLExtensions(glExtensions, "310 es", isFragment);
+				return result;
+
+			case "330":
+				var result = buildGLSLExtensions(glExtensions, "320 es", isFragment);
+				return result;
+
+			case "400":
+				return glExtensions;
+			case "410":
+				return glExtensions;
+			case "420":
+				return glExtensions;
+			case "430":
+				return glExtensions;
+			case "440":
+				return glExtensions;
+			case "450":
+				return glExtensions;
+			case "460":
+				return glExtensions;
+		}
```

Comment:
Pet-peeve, but I still don't get why this needs to be so overly complicated
```suggestion
		switch (glVersion)
		{
			default: 
				return glExtensions;

			case "300 es":
				var hasSeparateShaderObjects = false;
				for (extension in glExtensions)
				{
					if (extension.name == "GL_ARB_separate_shader_objects") hasSeparateShaderObjects = true;
					if (extension.name == "GL_EXT_separate_shader_objects") hasSeparateShaderObjects = true;
				}

				#if desktop
				if (!hasSeparateShaderObjects)
				{
					#if linux
					return glExtensions.concat([{name: "GL_EXT_separate_shader_objects", behavior: "require"}]);
					#else
					return glExtensions.concat([{name: "GL_ARB_separate_shader_objects", behavior: "require"}]);
					#end
				}
				#end
				return glExtensions;

			case "310 es":
				return buildGLSLExtensions(glExtensions, "300 es", isFragment);
			case "320 es":
				return buildGLSLExtensions(glExtensions, "310 es", isFragment);
			case "330":
				return buildGLSLExtensions(glExtensions, "320 es", isFragment);
		}
```

Reply:
this also seems functionally identical to 
```hx
switch (glVersion)
{
	default: 
		return glExtensions;

	#if desktop
	case "300 es", "310 es", "320 es", "330":
		var hasSeparateShaderObjects = false;
		for (extension in glExtensions)
		{
			if (extension.name == "GL_ARB_separate_shader_objects") hasSeparateShaderObjects = true;
			if (extension.name == "GL_EXT_separate_shader_objects") hasSeparateShaderObjects = true;
		}

		if (!hasSeparateShaderObjects)
		{
			#if linux
			return glExtensions.concat([{name: "GL_EXT_separate_shader_objects", behavior: "require"}]);
			#else
			return glExtensions.concat([{name: "GL_ARB_separate_shader_objects", behavior: "require"}]);
			#end
		}
		return glExtensions;
	#end
}
```

Reply:
I think it was only worded like this as, while testing, I anticipated requiring more version-specific changes, but ended up only needing a couple. I will definitely rewrite this to match your proposed syntax.